ARCHITECTURE.md

```diff
@@ -0,0 +1,456 @@
+# Architecture
+
+This document describes the high-level architecture of PyMC3.
+If you want to familiarize yourself with the code base, you are just in the right place!
+
+
+# Bird's Eye View
+
+TODO: Update diagram to include SMC
+![Architecture](docs/Architecture.png)
+
+On the highest level, PyMC3 lets you define probabilistic graphs or models
+that can be easily used to compute log probabilities for posterior inference
+or to draw random samples from priors or posteriors. 
+However there are numerous
+other useful sets of functionality, from prior and posterior predictive sampling,
+GraphviZ model, as well as various helper functions.
+
+PyMC3 includes a few inference techniques, in particular:
+* Markov chain Monte Carlo 
+* Variational Inference
+* Sequential Monte Carlo
+
+The most relevant modules of PyMC3 are shown visually below, as well as their
+relation to other major packages. Not all modules are shown, either because
+they are smaller or self explanatory in scope, or they're pending
+deprecation
+
+
+## Functionality not in PyMC3
+It is easier to start with functionality that is not present in PyMC3 but
+rather deferred to outside libraries. If seeking to understand any 
+of the topics below refer to that specific library
+
+### Aesara
+* Gradient computation
+* Random number generation
+* Low level tensor operation definition
+* Low level operation graphs
+
+### ArviZ
+* Plotting e.g. Trace plots, rank plots, posterior plots
+* MCMC sampling diagnostics e.g. Rhat, Effective Sample Size.
+* Model comparison, particularly efficient leave-one-out cross-validation approximation
+* Inference Data structure
+
+
+# Modules
+The codebase of PyMC3 is split among single Python file modules at the root
+level, as well as directories with Python code for logical groups of functionality.
+Admittedly the split between single `.py` module or directory is not defined by a strict
+criteria but tends to occur when single `.py` files would be "too big".
+We will with the modules needed for a "simple MCMC" model
+before detailing the remaining modules, such as Variational Inference, Ordinary Differential Equations,
+or Sequential Monte Carlo.
+
+## [model.py](https://docs.pymc.io/api/model.html#module-pymc3.model)
+This module contains pieces of code related to model definition and evaluation of the model.
+In no particular order they are
+
+* `ContextMeta`: The context manager that enables the `with pm.Model() as model` syntax
+* `Factor`: Defines the methods for the various logprobs for models
```

Comment:
Similar thing here

```suggestion
* {class}`~pymc3.Factor`: Defines the methods for the various logprobs for models
```

will generate a hyperlink to the right api page. See https://arviz-devs.github.io/arviz/contributing/developer_guide.html#hyperlinks for more details.